### DISCUSSION TITLE
docker: use multi-stage builds for smaller container sizes?

### DISCUSSION THREAD

**Discussion by stapelberg** [pretix/pretix#2745] [Ideas]
pretix 4.12’s docker container comes in at 649.69 MB compressed size. In `docker inspect pretix/standalone:stable`, I see a Size value of 1862 MB. This seems pretty big for my taste, and it would be nice to bring down the size for faster `docker push` and `docker pull`. I don’t know if docker startup or runtime are affected by fewer layers, but as far as I understand, reducing the size and number of layers is generally good :)

I looked into it a bit, and it seems to me like large parts of the docker container are left-overs from the build environment that are not actually required at runtime. One could even go so far and say that there should be an additional docker container that doesn’t contain a build environment at all, only a runtime environment.

This could be achieved by using docker’s multi-stage builds feature. The current Dockerfile would remain the first build stage, and then the required artifacts are copied over into a separate build that only needs the runtime dependencies, not the build dependencies.

I played around with this a bit, using python virtualenv to make for an easy-to-copy python environment. I got pretix to work far enough to do database migrations when starting up, but not far enough to deliver a non-error HTTP response. You can find my not-quite-working multi-stage dockerfile at https://gist.github.com/stapelberg/484992f3cb5946866277f9ff50c30ede. The resulting container comes in at 253 MB compressed size and 728 MB `docker inspect` Size.

Of course, to keep https://docs.pretix.eu/en/latest/admin/installation/docker_smallscale.html#install-a-plugin working, both build stages should be published — the current name would be the first build stage, and under a separate name the second build stage (for runtime) would be published.

**Comment by raphaelm** [comment]
Someone suggested this to me a few months ago, and I played around with it a little.

<hr>

For our SaaS deployment at pretix.eu, we use a different Dockerfile and I have a working draft that brings that docker image down from 4GB to 2GB using a multi-stage built. Still huge, but it shows that there is certainly something to gain.

The main low-hanging fruits for savings were (a) not having a nodejs in the final image (b) ability to delete some stuff from Debian we don't need like /usr/share/doc (c) removing some duplicate files such as larger static files in both the source as well as the build directory.

Still, there is a lot we can't get rid of. We use ``pdftk`` for some PDF manipulation operations, and debian provides pdftk through the ``pdftk-java`` implementation which means we need to include an entire JVM :roll_eyes:  Also, for a lot of system libraries (think ``libjpeg``) it's hard to tell for sure which ones some Python dependency might want to load at some point during runtime and which ones are safe to remove.

<hr>

The main reason I have not yet started experimenting with this for open-source pretix is the one you identified: Installing plugins. Installing plugins certainly needs many if not all of the build dependencies. So yes, what we could do is publish both docker images, as you suggested.

Multistage is obviously a big efficiency win if you have a location where you build and then have the resulting docker image as an artifact that you ship to other servers. This is the case for pretix.eu SaaS, but the vast majority of pretix installations are single-server installations. There are other people operating pretix on, say, Amazon Fargate clusters, but as far as I know, there are maybe 5-10 of these installations.

For a single-server installation, people will need to `docker pull` the fat image at some point if they want to install their first plugin. Our telemetry data suggests that 64% of active pretix installations with enabled telemetry have at least one plugin installed. So there might be another 36% out there who will benefit, although I think that number is actually lower because it is skewed by very recent installations who just have not installed any plugins *yet*.

<hr>

In case it's interesting at all, for our internal experiment I tried a three-part build, e.g.

```
FROM python:3.8 as baseimage

RUN # install everything needed in both build and runtime

FROM baseimage as builder

RUN # install everything only needed at build time
RUN # build steps

FROM baseimage as runtime

COPY # stuff
```

This mostly prevents some duplication of the long apt commands in the Dockerfile, and it guarantees e.g. the same python version in build and runtime. Not sure if there are any downsides to the approach.

**Reply by stapelberg** [reply]
Good to know you have this on your radar already! :)

> The main low-hanging fruits for savings were (a) not having a nodejs in the final image (b) ability to delete some stuff from Debian we don't need like /usr/share/doc (c) removing some duplicate files such as larger static files in both the source as well as the build directory.

That matches my impression. Instead of trimming down Debian, another thing folks like to do seems to be to use Alpine for their Docker containers, which is much smaller out of the box. I’m not sure if we want to go to such lengths, though.

> We use pdftk for some PDF manipulation operations, and debian provides pdftk through the pdftk-java implementation which means we need to include an entire JVM roll_eyes 

This got me curious. AFAICT, the only call to pdftk is this one: https://github.com/pretix/pretix/blob/69b482849a6bd1f89f47db7cea75be09f420cbe2/src/pretix/base/pdf.py#L888-L925

…and that one *already* has an alternative pure-python implementation. Any reason to stick with pdftk instead of just using PyPDF2 always?

> Also, for a lot of system libraries (think libjpeg) it's hard to tell for sure which ones some Python dependency might want to load at some point during runtime and which ones are safe to remove.

Yeah, that is certainly a worry of mine. It’ll require careful testing to make sure all the features still work.

> Multistage is obviously a big efficiency win if you have a location where you build and then have the resulting docker image as an artifact that you ship to other servers. This is the case for pretix.eu SaaS, but the vast majority of pretix installations are single-server installations. 

This is not a contradiction. My single-server workflow involves pulling+building+pushing new docker images on my workstation, and then only pulling+running the result on my server. So, yes, I need to stomach a heavy docker image, but only on my well-equipped workstation. If there was a way for me to do a multi-stage build, where the build stage comes from pretix upstream, and the run stage is upstream + my plugins, I could still benefit from smaller images where it matters (on my single server).

> This mostly prevents some duplication of the long apt commands in the Dockerfile, and it guarantees e.g. the same python version in build and runtime. Not sure if there are any downsides to the approach.

I would assume that it results in more layers than a 2-stage build, but probably that’s fine to ignore, as reducing duplication seems more important to me.

**Reply by raphaelm** [reply]
> That matches my impression. Instead of trimming down Debian, another thing folks like to do seems to be to use Alpine for their Docker containers, which is much smaller out of the box. I’m not sure if we want to go to such lengths, though.

Yeah, switching to Alpine would be a lot of work, and there's no guarantee it will be any better. Alpine is great for e.g. wrapping statically linked binaries in a Docker image, which is why Alpine has become so popular, but with a language like Python and Python specifically, it can also go the other way.

Here's an interesting blog post on that:
https://pythonspeed.com/articles/alpine-docker-python/

tl;dr: Alpine ist based on musl instead of glibc, hence it can't use precompiled libraries from PyPI, thus requiring to recompile all python dependencies that are not packaged by Alpine itself. Also, there can be runtime issues because [CPython wasn't created with musl in mind](https://bugs.python.org/issue32307).

> Any reason to stick with pdftk instead of just using PyPDF2 always?

As the git history confirms, the pdftk codepath was added **after** the PyPDF2 codepath so there must have been a compelling reason to do so. Unfortunately, past-me did not note the reason down in 2018. It could have been performance, but it could also have been a compatibility thing. The background PDF file is user-uploaded and PDF files can get weird. Maybe pdftk was better at dealing with some weird files, or keeping CMYK color spaces intact for printing, or something like that. Would need some testing to figure out if there's a big difference that's still reproducible, but there probably was.

> This is not a contradiction. My single-server workflow involves pulling+building+pushing new docker images on my workstation, and then only pulling+running the result on my server. So, yes, I need to stomach a heavy docker image, but only on my well-equipped workstation. If there was a way for me to do a multi-stage build, where the build stage comes from pretix upstream, and the run stage is upstream + my plugins, I could still benefit from smaller images where it matters (on my single server).

Good point!

**Reply by raphaelm** [reply]
> This is not a contradiction. My single-server workflow involves pulling+building+pushing new docker images on my workstation, and then only pulling+running the result on my server. So, yes, I need to stomach a heavy docker image, but only on my well-equipped workstation. If there was a way for me to do a multi-stage build, where the build stage comes from pretix upstream, and the run stage is upstream + my plugins, I could still benefit from smaller images where it matters (on my single server).

Ah, however, there's a big counterargument as well: Your customized Dockerfile that installs your plugins would need to start at the image we publish that includes the build tools. So you need to lots of the ``COPY`` commands into your customized Dockerfile, or am I missing something?

Currently, customized Dockerfiles consist of little more than one additional line of ``pip install``. With multi-stage build, they would get much more complicated, and people might need to update them more often if something about the toolchain changes.

**Reply by stapelberg** [reply]
> Alpine is great for e.g. wrapping statically linked binaries in a Docker image, which is why Alpine has become so popular, but with a language like Python and Python specifically, it can also go the other way.
> 
> Here's an interesting blog post on that:
https://pythonspeed.com/articles/alpine-docker-python/

Ah, yes, I remember that post! Curious why so many tutorials suggest using alpine with django apps, though. Maybe enough software just works that many folks never run into problems?

> It could have been performance, but it could also have been a compatibility thing. The background PDF file is user-uploaded and PDF files can get weird. Maybe pdftk was better at dealing with some weird files, or keeping CMYK color spaces intact for printing, or something like that. Would need some testing to figure out if there's a big difference that's still reproducible, but there probably was.

Okay. FWIW, I don’t use the background PDF file feature at all, so for my needs I would gladly just not install pdftk :)

> Ah, however, there's a big counterargument as well: Your customized Dockerfile that installs your plugins would need to start at the image we publish that includes the build tools. So you need to lots of the COPY commands into your customized Dockerfile, or am I missing something?

Yes, you’re correct. I would need to copy&paste, and potentially update, your last-stage build instructions. Unless I’m missing some Docker feature, of course :) 

Keep in mind that this is optional — people who don’t care about multi-stage builds would be exactly as well off as before, and those who do want to pay the maintenance/setup cost for getting multi-stage builds with smaller outputs can chose to do so.
